rfc/010-distributed-coordinator.md

```diff
@@ -0,0 +1,249 @@
+
+* A successful [Peer recovery](#peer-recovery).
+* A successful recovery through the `userapi.Recover` RPC.
+
+### Kubernetes changes
+
+#### Coordinator service
+
+Coordinators are going to recover themselves from peers that are already recovered (or set).
+This means that we need to be able to distinguish Coordinators that have secrets from Coordinators that have none.
+Since we're running in Kubernetes, we can leverage the built-in probe types.
+
+We add a few new paths to the existing HTTP server for metrics, supporting the following checks (returning status code 503 on failure):
+
+* `/probe/startup`: returns 200 when all ports are serving and [peer recovery](#peer-recovery) was attempted once.
+* `/probe/liveness`: returns 200 unless the Coordinator is recovered but can't read the transaction store.
+* `/probe/readiness`: returns 200 if the Coordinator has an active manifest and isn't in recovery mode.
+
+Using these probes, we expose the Coordinators in different ways, depending on audience:
+
+* A `cooordinator-ready` service backed by ready Coordinators should be used by initializers and verifying clients.
+* The `coordinator` service now accepts unready endpoints (`publishNotReadyAddresses=true`) and should be used for `set`.
+  The idea here is backwards compatibility with documented workflows, see [Alternatives considered](#alternatives-considered).
+
+#### Peer discovery
+
+A recovering Coordinator needs to know potential peers to recover from.
+We add a background watcher for k8s pods that monitors existing coordinator pods, filters for those that are ready and populates a local lookup table.
+Such functionality is core to projects like operator-sdk and kubebuilder.
+However, those are very opinionated and rather monolithic, so it may be better to implement this using the watch primitives in `client-go`.
+The implementation could probably share some code with `KubeClient.WaitFor`.
+
+#### Coordinator pods
+
+The Coordinator pods should have a soft anti-affinity[^3] to themselves.
+
+[^3]: <https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#inter-pod-affinity-and-anti-affinity>
+
+### Manifest changes
+
+In order to allow Coordinators to recover from their peers, we need to define which workloads are allowed to recover.
+There are two reasons why we want to make this explicit:
+
+1. Allowing a Coordinator with a different policy to recover is a prerequisite for automatic updates.
+2. The Coordinator policy can't be embedded into the Coordinator during build, and deriving it at runtime is inconvenient.
+
+Thus, we introduce a new field to the manifest that specifies roles available to the verified identity.
+For now, the only known role is `coordinator`, but this could be extended in the future (for example: delegate CAs).
+
+```golang
+type PolicyEntry struct {
+  SANs             []string
+  WorkloadSecretID string `json:",omitempty"`
+  Roles            []string `json:",omitempty"` // <-- New in RFC009.
```

Comment:
```suggestion
  Roles            []string `json:",omitempty"` // <-- New in RFC009.
```
I know this is probably only an example, but we'll want to type the roles, I think?

Reply:
I don't understand the suggested diff, but I think you meant to say roles should be an enum array, not a string array? If only Contrast is the audience for these I'd agree, but we could also make them available to users as certificate extensions, in which case we should probably not restrict ourselves here.